controllers/vmware/virtualmachinegroup_controller.go

```diff
@@ -0,0 +1,138 @@
+	predicateLog := ctrl.LoggerFrom(ctx).WithValues("controller", "virtualmachinegroup")
+
+	reconciler := &VirtualMachineGroupReconciler{
+		Client:   controllerManagerCtx.Client,
+		Recorder: mgr.GetEventRecorderFor("virtualmachinegroup-controller"),
+	}
+
+	// Predicate: only allow VMG with the cluster-name label. Ensures the controller only works on VMG objects created by CAPV.
+	hasClusterNameLabel := predicate.NewPredicateFuncs(func(obj ctrlclient.Object) bool {
+		labels := obj.GetLabels()
+		if labels == nil {
+			return false
+		}
+		_, ok := labels[clusterv1.ClusterNameLabel]
+		return ok
+	})
+
+	builder := ctrl.NewControllerManagedBy(mgr).
+		For(&vmoprv1.VirtualMachineGroup{}).
+		WithOptions(options).
+		WithEventFilter(hasClusterNameLabel).
+		Watches(
+			&clusterv1.Cluster{},
+			handler.EnqueueRequestsFromMapFunc(reconciler.ClusterToVirtualMachineGroup),
+		).
+		Watches(
+			&vmwarev1.VSphereMachine{},
+			handler.EnqueueRequestsFromMapFunc(reconciler.VSphereMachineToVirtualMachineGroup),
+			ctrlbldr.WithPredicates(
+				predicate.Funcs{
+					UpdateFunc:  func(event.UpdateEvent) bool { return false },
+					CreateFunc:  func(event.CreateEvent) bool { return true },
+					DeleteFunc:  func(event.DeleteEvent) bool { return true },
+					GenericFunc: func(event.GenericEvent) bool { return false },
+				}),
+		).
+		WithEventFilter(predicates.ResourceHasFilterLabel(mgr.GetScheme(), predicateLog, controllerManagerCtx.WatchFilterValue))
+
+	return builder.Complete(reconciler)
+}
+
+// ClusterToVirtualMachineGroup maps Cluster events to VirtualMachineGroup reconcile requests.
+func (r *VirtualMachineGroupReconciler) ClusterToVirtualMachineGroup(_ context.Context, a ctrlclient.Object) []reconcile.Request {
+	cluster, ok := a.(*clusterv1.Cluster)
+	if !ok {
+		return nil
+	}
+
+	// Always enqueue a request for the "would-be VMG"
+	return []reconcile.Request{{
+		NamespacedName: apitypes.NamespacedName{
+			Namespace: cluster.Namespace,
+			Name:      cluster.Name,
+		},
+	}}
+}
+
+// VSphereMachineToVirtualMachineGroup maps VSphereMachine events to VirtualMachineGroup reconcile requests.
+// This handler only processes VSphereMachine objects for Day-2 operations when VMG could be found, ensuring
+// VMG member list in sync with VSphereMachines. If no corresponding VMG is found, this is a no-op.
+func (r *VirtualMachineGroupReconciler) VSphereMachineToVirtualMachineGroup(ctx context.Context, a ctrlclient.Object) []reconcile.Request {
+	vSphereMachine, ok := a.(*vmwarev1.VSphereMachine)
+	if !ok {
+		return nil
+	}
+
+	clusterName, ok := vSphereMachine.Labels[clusterv1.ClusterNameLabel]
+	if !ok || clusterName == "" {
+		return nil
+	}
+
+	vmg := &vmoprv1.VirtualMachineGroup{}
+	err := r.Client.Get(ctx, apitypes.NamespacedName{
+		Namespace: vSphereMachine.Namespace,
+		Name:      clusterName,
+	}, vmg)
+
+	if err != nil {
+		return nil
+	}
```

Comment:
We'd better add comments to explain the whole logic of the controller. It 's a workaround approach after all.